### ISSUE TITLE
Allow skipping of specific crates in 'cargo doc'

### ISSUE THREAD

**Issue by LPGhatguy** [rust-lang/cargo#4049]
When building projects on Windows that use `winapi` somewhere in a downstream dependency, generating documentation takes a long time. Usually in these projects, I don't actually care about `winapi` types; the extra 20 minutes before I can use the documentation I _do_ care about is a drag!

Does it make sense to be able to skip documentation of a specific crate, like `cargo doc --skip-package winapi`? Any references to the types re-exported by crates I consume directly would have issues, I assume.

**Comment by alexcrichton** [comment]
Perhaps https://github.com/rust-lang/cargo/pull/4031 would help here?

**Comment by carols10cents** [comment]
I don't think `--exclude` helps here, since `--exclude` only works with `--all` and `--all` only works with workspaces.

From a non-workspace crate that lists `rand` as a dependency:

```
$ cargo doc --exclude rand
error: Unknown flag: '--exclude'

Usage:
    cargo doc [options]

$ cargo doc --all --exclude rand
error: Unknown flag: '--exclude'

Usage:
    cargo doc [options]
```

**Comment by icefoxen** [comment]
I second this; this is troublesome for gfx-rs, which has many sub-crates which only work on certain platforms.  Trying to even run `cargo doc` on it will always try to build at least one thing which can't build on any platform.

**Comment by searing** [comment]
This would be great - trying to document even simple crates can easily take a very long time.

I think the ideal solution would be a way to (optionally) externally link to docs.rs for crates.io dependencies. But, barring that, simply not linking anywhere would be a nice feature. I really don't need to generate a local copy of Diesel's documentation for a simple one-file library that uses it, for example.

**Comment by jesskfullwood** [comment]
Worth noting that crates sometime fail to compile the docs while otherwise working fine. This is currently the case with `diesel` on `nightly`. I'm unable to generate docs for my own project at the moment due to the `diesel` dependency

**Comment by brokenthorn** [comment]
What about a `--project-only` flag? (maybe use [docs.rs](https://docs.rs) for external links)

**Comment by cpick** [comment]
The original request was motivated by wanting to decrease doc-build times, but some of the additional use cases (including my own) are due to wanting documentation for other dependencies even if one fails to build.

This case would certainly be well served by an ability to manually skip the problematic crate(s), but also by a `--persist-through-errors` option if that could be made to continue generating the rest of the documentation.

If that sounds possible (or easier than `--skip-crate`) I could open a new issues to request that option?

**Comment by thkien** [comment]
I'm trying to find a solution for similar issue then came across this old ticket. 

According to "The Cargo Book", `cargo doc` has following options:

- `--no-deps`: Do not build documentation for dependencies ... ← this is the one that solved my issue
- `-p/--package spec...`: Document only the specified packages ...
- `--exclude SPEC...`: Exclude the specified packages ...

So probably this issue should be fixed (?)

**Comment by melMass** [comment]
> So probably this issue should be fixed (?)

Indeed, and the highlighted answer has nothing to do with the issue 😅

**Comment by Sunshine40** [comment]
> According to "The Cargo Book", `cargo doc` has following options:
> 
> * `--no-deps`: Do not build documentation for dependencies ... ← this is the one that solved my issue

There is one problem: if `package-b` depends on `package-a` in the same workspace, then with `--workspace --no-deps`, everything seems fine except that the links to items in `package-a` are no longer available in `package-b`'s doc pages.

Ideally, we need specific packages to be hidden in documentation, but without making package docs generated with a single command isolated from each other which is how `--no-deps` currently works in order not to bring up dependencies.

**Comment by ChrisDenton** [comment]
I'll note here that the [link above](https://github.com/microsoft/windows-rs/issues/2665#issuecomment-1739671565) shows this is relevant to the `windows` and `windows-sys` crates too (successors to the `winapi` crate). These are monolithic crates so will necessarily generate a lot of docs.

**Comment by epage** [comment]
> The original request was motivated by wanting to decrease doc-build times, but some of the additional use cases (including my own) are due to wanting documentation for other dependencies even if one fails to build.
> 
> This case would certainly be well served by an ability to manually skip the problematic crate(s), but also by a `--persist-through-errors` option if that could be made to continue generating the rest of the documentation.
> 
> If that sounds possible (or easier than `--skip-crate`) I could open a new issues to request that option?

`cargo doc` got a `--keep-going` flag in #12568.  Might not solve all problems but might help with some.

**Comment by epage** [comment]
Also, would #2025 reduce the impact of this?

**Comment by ChrisDenton** [comment]
cc @LikeLakers2

**Comment by LikeLakers2** [comment]
> [@epage] Also, would #2025 reduce the impact of this?

In the case I described at <https://github.com/microsoft/windows-rs/issues/2665>? Yes, I believe it would, as it would keep the monolithic `windows-sys` and `winapi` crates from being documented locally unless I'm using them directly.

**Comment by ehuss** [comment]
The cargo team discussed this a couple weeks ago and decided to close this issue. There were a few other alternatives that we were considering:
* [public-private dependencies](https://github.com/rust-lang/cargo/issues/6129) could provide a way to avoid generating documentation for unreachable private dependencies. Though it's not clear to me what the design considerations would be there (how that would be enabled, whether it would still include direct private dependencies, etc.). Depending on how the design considerations for that change, we may want to re-open this issue if it doesn't fully address this use case.
* [rustdoc map](https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#rustdoc-map) would help with the problem that documenting all dependencies is expensive by instead directing links to docs.rs. Though by itself that has the downsides that it requires an internet connection, and the navigation may not be as good (crates listed in the sidebar). In theory you could construct your own local docs collection to point to, but that's probably beyond the scope here.
* https://github.com/rust-lang/cargo/issues/2025 is a related issue that could reduce the amount of crates documented.

**Comment by Hawk777** [comment]
@ehuss I actually have a use case for this exact request, that would *not* AFAICT be covered by any of the three things you linked to.

I am working on a project which depends on the `gtk4` crate, as well as a handful of other crates. Ideally I would like all my documentation to be together in one browser tab (both the docs for my own project and the docs for my direct dependencies that I’m using to develop the project¹). However, for license reasons, the `gtk4` crate doesn’t actually include any documentation. There is a somewhat complicated workaround to include it, but it apparently involves fetching the sources locally, installing a Cargo plugin, running a Python script, and then using a nightly Rustdoc, so it’s a big hurdle. Instead, I open separate browser tabs and use the web versions of the docs for those crates. Therefore, I would prefer to exclude `gtk4` from my local docs (because it’s just useless clutter, all the type and function names are there but with no actual documentation attached, and there are a *lot* of symbols cluttering up e.g. search results). Public/private and only-direct would not solve this because `gtk4` *is* a direct dependency of my project, but I want to exclude it from docs. `--no-deps` doesn’t solve this because I *do* want to include all my other dependencies (at least the direct ones). Maps I don’t think would help because it looks like it only allows you to redirect links on a per-registry basis, whereas `gtk4` is fetched from `crates.io` but *not* documented (usefully) on `docs.rs` (the useful docs live on `gtk-rs.org` instead).

Does this use case warrant reopening this ticket?

¹The public/private or only-direct features would independently be nice to reduce the amount of clutter from crates I don’t directly depend on, but are not the point of this comment.

**Comment by epage** [comment]
Thank you for sharing your use case!

Speaking only for myself, that seems like a fairly niche workflow to design an entire feature around.

**Comment by Hawk777** [comment]
You’re right, it is a bit obscure. Figured I’d share it in case it influenced anything, but I understand if it doesn’t.

**Comment by epage** [comment]
I agree it is important to share. As we collect more use cases and as features evolve, we can look across the use cases to either go ahead with a past idea or come up with something new.
